ce/yaml/yaml-types.ts

```diff
@@ -0,0 +1,189 @@
+import { ValidationError } from '../interfaces/validation-error';
+import { isNullish } from '../util/checks';
+
+export class YAMLDictionary extends YAMLMap<string, any> { }
+export class YAMLSequence extends YAMLSeq<any> { }
+export class YAMLScalar extends Scalar<any> { }
+export type Primitive = string | number | boolean;
+export type Node = YAMLDictionary | YAMLSequence | YAMLScalar;
+export type Range = [number, number, number];
+
+export /** @internal */ abstract class Yaml<ThisType extends Node = Node> {
+  constructor(/** @internal */ node?: ThisType, protected parent?: Yaml<Node>, protected key?: string) {
+    this.node = node;
+    if (!(<NodeFactory<ThisType>>(this.constructor)).create) {
+      throw new Error(`class ${this.constructor.name} is missing implementation for create`);
+    }
+  }
+
+  get root(): Yaml {
+    return this.parent ? this.parent.root : this;
+  }
+
+  protected createNode(): ThisType {
+    return (<NodeFactory<ThisType>>this.constructor).create();
+  }
+
+  /**@internal*/ static create() {
+    throw new Error('creator not Not implemented on base class.');
+  }
+
+  private _node: ThisType | undefined;
+
+  get node(): ThisType | undefined {
+    if (this._node) {
+      return this._node;
+    }
+
+    if (this.key && isMap(this.parent?.node)) {
+      this._node = <any>this.parent!.node.get(this.key, true);
+    }
+
+    return this._node;
+  }
+
+  set node(n: ThisType | undefined) {
+    this._node = n;
+  }
+
+  sourcePosition(key?: string | number): Range | undefined {
+    if (!this.node) {
+      return undefined;
+    }
+    if (key !== undefined) {
+      if ((isMap(this.node) || isSeq(this.node))) {
+        const node = <Node>this.node.get(<any>key);
+        if (node) {
+          return node.range || undefined;
+        }
+        return undefined;
+      }
+      if (isScalar(this.node)) {
+        throw new Error('Scalar does not have a key to get a source position');
+      }
+    }
+    return this.node?.range || undefined;
+  }
+
+  /** will dispose of this object if it is empty (or forced) */
+  dispose(force = false, deleteFromParent = true) {
+    if ((this.empty || force) && this.node) {
+      if (deleteFromParent) {
+        this.parent?.deleteChild(this);
+      }
+      this.node = undefined;
+    }
+  }
+
+  /** if this node has any data, this should return false */
+  get empty(): boolean {
+    if (isCollection(this.node)) {
+      return !(this.node?.items.length);
+    }
+    return !isNullish(this.node?.value);
+  }
+  /** @internal */ get exists(): boolean {
+    if (this.node) {
+      return true;
+    }
+    // well, if we're lazy and haven't instantiated it yet, check if it's created.
+    if (this.key && isMap(this.parent?.node)) {
+      this.node = <any>this.parent!.node.get(this.key);
+      if (this.node) {
+        return true;
+      }
+    }
+    return false;
+  }
+  /** @internal */ assert(recreateIfDisposed = false, node = this.node) {
```

Comment:
Can this function by annotated with `asserts this is { node: ThisType }` to avoid all the exclamation points in callers?

Reply:
Such was my original intent, but I've been dancing around a couple of errors (narrowing down to when I tried to use it with `this.parent.assert(...)`  -- the compiler is real tetchy about it giving me with `Assertions require every name in the call target to be declared with an explicit type annotation.ts(2775)` ... regardless of how much I try to be explicit.

but, after rabbit-holing for an hour or two, I came up with 
``` ts
 assert(recreateIfDisposed = false, node = this.node): asserts this is Yaml<ThisType> & { node: ThisType } { ... } 
```

And just used `(<any>this.parent).assert(true);` and `this.parent.node!.` in the one place I needed to. This did drastically cut down on my explicit assertions.